src/devices/oregon_scientific.c

```diff
@@ -104,15 +104,15 @@ static unsigned short int cm180i_power(uint8_t const *msg, unsigned int offset)
     return val;
 }
 
-static unsigned long long cm180i_total(uint8_t const *msg)
+static uint64_t cm180i_total(uint8_t const *msg)
 {
-    unsigned long long val = 0;
+    uint64_t val = 0;
     if ((msg[1] & 0x0F) == 0) {
         // Sensor returns total only if nibble#4 == 0
-        val = (unsigned long long)msg[14] << 40;
-        val += (unsigned long long)msg[13] << 32;
-        val += (unsigned long)msg[12] << 24;
-        val += (unsigned long)msg[11] << 16;
+        val = (uint64_t)msg[14] << 40;
+        val += (uint64_t)msg[13] << 32;
+        val += msg[12] << 24;
```

Comment:
I understood the policy to assume "implicit integer promotion" where integer is a 32 bit and only write (u)int32_t when it HAS to be.

IMHO, it does not "have to be" uint32_t as we do not care about the exact size in memory and no bits will be shifted outside the 32bit range and integer promotion is supposed to ensure 32bit and msg[12] is already unsigned.

If it is needed here then it is also needed for "val += msg[11] << 16"

Also see #1193 .


My personal policy is to put (u)int16_t, (u)int32_t, etc. everywhere where I need more than 8 bit as on a microcontroller that is less than 32bit, int is only as big as the microcontroller's native register size.  And I use (u)int8_t as well to help ensure that I use as little memory as possible on microcontrollers.


Please confirm that the policy requires to cast this line, and if it is or not needed for the next line as well.

Reply:
Not about policy: `<<24` will fill an int and thus might get negative -- this might be ok with bit operations but not with `+=`

Reply:
It's worse than I thought:
```c
#include <stdio.h>
#include <inttypes.h>

int main() {
   uint8_t m[32];
   uint64_t sum=0;
   uint64_t or_op=0;
   uint64_t or_op_32cast=0;

   m[12]=0xCF;

   printf("%08X << 24 is %08X\n", m[12], m[12]<<24);
   sum += m[12]<<24;
   or_op |= m[12]<<24;
   or_op_32cast |= (uint32_t)m[12]<<24;
   printf("sum      %016" PRIX64 "\n", sum);
   printf(" or      %016" PRIX64 "\n", or_op);
   printf(" or cast %016" PRIX64 "\n", or_op_32cast);
   return 0;
}
```
```plaintext
000000CF << 24 is CF000000
sum      FFFFFFFFCF000000
 or      FFFFFFFFCF000000
 or cast 00000000CF000000
```

Which pleads for casting as strict as possible rather than relying on interpretations of what the compiler does by default.

Will update the code later.

Reply:
I fixed the relevant commit, rebased and forced pushed.

I also changed https://github.com/merbanan/rtl_433/blob/dcfd6cc43c4998cbca3fa232226f81965de63987/src/devices/oregon_scientific.c#L145 .

Reply:
The compiler isn't funny here, it's by spec for integer promotion and sign extend.
No need to cast on anything `<<23` or lower, we can't deal with `sizeof int < 32` anyway.